lib/SILOptimizer/Utils/Existential.cpp

```diff
@@ -309,3 +311,77 @@ ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &openedUse) {
   }
   assert(isValid());
 }
+
+/// Initialize a ConcreteExistentialInfo based on the already computed concrete
+/// type and protocol declaration. It determines the OpenedArchetypeDef
+/// and SubstituionMap for the ArgOperand argument.
+/// We need the OpenedArchetypeDef to be able to cast it to the concrete type.
+/// findInitExistential helps us determine this OpenedArchetypeDef. For cases
+/// where OpenedArchetypeDef can not be found from findInitExistential (because
+/// there was no InitExistential), then we do
+/// extra effort in trying to find an OpenedArchetypeDef for AllocStackInst
+/// using getAddressOfStackInit.
+ConcreteExistentialInfo::ConcreteExistentialInfo(Operand &ArgOperand,
+                                                 CanType ConcreteTy,
+                                                 ProtocolDecl *Protocol)
+    : ConcreteExistentialInfo(ArgOperand) {
+
+  // If we found an InitExistential, assert that ConcreteType we determined is
+  // same as ConcreteTy argument.
+  if (InitExistential) {
+    assert(ConcreteType == ConcreteTy);
+    assert(isValid());
+    return;
+  }
+
+  ConcreteType = ConcreteTy;
+  OpenedArchetypeDef = ArgOperand.get();
+
+  // If findInitExistential call from the other constructor did not update the
+  // OpenedArchetypeDef (because it did not find an InitExistential) and that
+  // the original Arg is an alloc_stack instruction, then we determine the
+  // OpenedArchetypeDef using getAddressOfStackInit. Please keep in mind that an
+  // alloc_stack can be an argument to apply (which could have no associated
+  // InitExistential), thus we need to determine the OpenedArchetypeDef for it.
+  // This is the extra effort.
+  SILInstruction *User = ArgOperand.getUser();
+  SILModule &M = User->getModule();
+  if (auto *ASI = dyn_cast<AllocStackInst>(OpenedArchetypeDef)) {
+    bool copied = false;
+    if (SILValue Src = getAddressOfStackInit(ASI, User, copied)) {
+      OpenedArchetypeDef = Src;
+    }
+    isCopied = copied;
+  }
+
+  // Bail, if we did not find an opened existential.
+  if (!(isa<OpenExistentialRefInst>(OpenedArchetypeDef) ||
+        isa<OpenExistentialAddrInst>(OpenedArchetypeDef)))
+    return;
+
+  // We have the open_existential; we still need the conformance.
+  auto ConformanceRef =
+      M.getSwiftModule()->lookupConformance(ConcreteType, Protocol);
+  if (!ConformanceRef)
+    return;
+
+  // Assert that the conformance is complete.
+  auto *ConcreteConformance = ConformanceRef.getValue().getConcrete();
+  assert(ConcreteConformance->isComplete());
+
+  /// Determine the ExistentialConformances and SubstitutionMap.
+  auto ExistentialConformances = ArrayRef<ProtocolConformanceRef>(
+      ProtocolConformanceRef(ConcreteConformance));
```

Comment:
Isn't this a use-after-free? I would rely on the implicit conversion between `T[]` and `ArrayRef<T>` and just pass `{ProtocolConformanceRef(ConcreteConformance)}` as the parameter to SubstitutionMap::get below

Reply:
I tried this before and even now. I get the following error:
```
../Existential.cpp:378:21: error: call to 'get' is ambiguous
  ExistentialSubs = SubstitutionMap::get(ExistentialSig, {ConcreteType},
                    ^~~~~~~~~~~~~~~~~~~~
.../swift/include/swift/AST/SubstitutionMap.h:105:26: note: candidate function
  static SubstitutionMap get(GenericSignature *genericSig,
                         ^.../swift/include/swift/AST/SubstitutionMap.h:118:26: note: candidate function
  static SubstitutionMap get(GenericSignature *genericSig,
                         ^
1 error generated.
```
Code changed:

```
ExistentialSubs = SubstitutionMap::get(ExistentialSig, {ConcreteType},
                                         /*ExistentialConformances*/ {ProtocolConformanceRef(ConcreteConformance)});
```